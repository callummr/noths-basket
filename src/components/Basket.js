import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const Container = styled.aside`
  order: -1;
  width: 100%;
  margin-bottom: 20px;
  @media (min-width: 768px) {
    order: 1;
    width: 250px;
  }
  background-color: #fff;
  padding: 15px;
`

const Title = styled.h1`
  color: #333333;
  font-size: 30px;
  margin-bottom: 20px;
`

const ProductTable = styled.table`
  margin-bottom: 20px;
  width: 100%;

  th {
    font-weight: 600;
    text-align: left;
    padding-bottom: 10px;
  }
`

const RemoveButton = styled.button`
  color: red;
  cursor: pointer;
  font-weight: bold;
  border: none;
  font-size: 1.2em;
`

const getTotalPrice = (basket, products) =>
  Object.entries(basket)
    .map(([id, count]) => products.find(p => p.id === id).price * count)
    .reduce((prev, next) => prev + next, 0)
    .toFixed(2)

const BasketItems = ({ products, basket, onRemove }) => {
  const items = Object.entries(basket).map(([id, count]) => ({
    id,
    count,
    name: products.find(p => p.id === id).name,
  }))

  return items.map(item => (
    <tr key={item.id}>
      <td>{item.name}</td>
      <td>{item.count}</td>
      <td>
        <RemoveButton onClick={() => onRemove(item.id)}>X</RemoveButton>
      </td>
    </tr>
  ))
}

const Basket = props => (
  <Container>
    <Title>Basket</Title>

    <ProductTable>
      <thead>
        <tr>
          <th>Product name</th>
          <th>qty</th>
          <th />
        </tr>
      </thead>

      <tbody>
        <BasketItems
          products={props.products}
          basket={props.basket}
          onRemove={props.onRemove}
        />
      </tbody>
    </ProductTable>

    <div>
      <span>Total: </span>
      <span>Â£{getTotalPrice(props.basket, props.products)}</span>
    </div>
  </Container>
)

Basket.propTypes = {
  products: PropTypes.arrayOf(PropTypes.object).isRequired,
  basket: PropTypes.objectOf(PropTypes.number).isRequired,
  onRemove: PropTypes.func.isRequired,
}

export default Basket
