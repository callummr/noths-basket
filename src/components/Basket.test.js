import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import Basket from './Basket'

const basketProps = {
  products: [
    { id: '001', name: 'test-name', price: 1.5 },
    { id: '002', name: 'test-name-2', price: 3.95 },
  ],
  basket: { '001': 2 },
  onRemove: jest.fn(),
}

describe('Basket', () => {
  it('matches snapshot', () => {
    const instance = <Basket {...basketProps} />
    expect(renderer.create(instance).toJSON()).toMatchSnapshot()
  })

  it('calls onRemove with product id when clicking remove', () => {
    const wrapper = mount(<Basket {...basketProps} />)
    wrapper
      .find('button')
      .first()
      .simulate('click')

    expect(basketProps.onRemove).toHaveBeenCalled()
  })
})
