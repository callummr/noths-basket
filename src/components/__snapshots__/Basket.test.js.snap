// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basket matches snapshot 1`] = `
<aside
  className="sc-bdVaJa uWOdI"
>
  <h1
    className="sc-bwzfXH kTbnZQ"
  >
    Basket
  </h1>
  <table
    className="sc-htpNat lfvxlQ"
  >
    <thead>
      <tr>
        <th>
          Product name
        </th>
        <th>
          qty
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          test-name
        </td>
        <td>
          2
        </td>
        <td>
          <button
            className="sc-bxivhb jMmaYx"
            onClick={[Function]}
          >
            X
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <span>
      Total: 
    </span>
    <span>
      Â£
      3.00
    </span>
  </div>
</aside>
`;
